### PR TITLE
Convert to Typescript

### PR DESCRIPTION
**Pull request by davidgoli**
Addresses #229 

Note there is a breaking change here: the types exported from `dist/es6/index.d.ts` do not conform to the structure in the original `index.d.ts`. Using `RRule` as both a namespace and a class seemed overly complicated, so I simplified it, but open to discussion.

I'd say this PR probably warrants a minor version bump in the package.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,22 +17,41 @@
     "rfc"
   ],
   "author": "Jakub Roztocil and Lars Schöning",
-  "main": "dist/index",
+  "main": "dist/es5",
```
Comment (nickarora) [inline package.json]: 💛 I believe `bower.json` will need to be similarly updated to point to the correct distribution

File: README.md [hunk 2]
```diff
@@ -43,18 +43,16 @@ Alternatively, download
 $ yarn add rrule
 ```
 
-```javascript
-var RRule = require('rrule').RRule
-var RRuleSet = require('rrule').RRuleSet
-var rrulestr = require('rrule').rrulestr
+```es6
+import { RRule, RRuleSet, rrulestr } from 'rrule/es6'
```
Comment (nickarora) [inline README.md]: 💛 I don't think its mentioned anywhere in the README but we should point out that you can use

```
import { RRule, RRuleSet, rrulestr } from 'rrule'
```

to pull in RRule that requires no transpiling

File: src/helpers.ts [hunk 3]
```diff
@@ -15,12 +15,12 @@ const range = function (start, end) {
   return rang
 }
 
-const repeat = function (value, times) {
+const repeat = function<T>(value: T | T[], times: number): T[] | T[][] {
```
Comment (mikelovesrobots) [inline src/helpers.ts]: This must have been "fun" to convert to typescript.

File: src/rrule.ts [hunk 4]
```diff
@@ -21,59 +14,186 @@ import {
...
+        between: []
+      }
 
     // used by toString()
     this.origOptions = {}
 
     this.options = {}
 
-    const invalid = []
+    const invalid: any[] = []
```
Comment (SamRaymer) [inline src/rrule.ts]: Could we use `unknown` and some type guards instead of `any` here?

File: src/rrule.ts [hunk 5]
```diff
@@ -277,29 +407,36 @@ export default class RRule {
...
-              if (/^[+-]?\d+$/.test(value[j])) value[j] = Number(value[j])
+              if (/^[+-]?\d+$/.test(value[j].toString())) {
+                value[j] = Number(value[j])
+              }
             }
           } else if (/^[+-]?\d+$/.test(value)) {
             value = Number(value)
           }
           key = key.toLowerCase()
+          // @ts-ignore
```
Comment (SamRaymer) [inline src/rrule.ts]: I get really intensely curious when I see `@ts-ignore`'s, could we leave little notes about what we're avoiding with them?

File: src/rrule.ts [hunk 6]
```diff
@@ -720,6 +908,7 @@ export default class RRule {
             // XXX: can this ever be in the array?
             // - compare the actual date instead?
             if (!contains(poslist, res)) poslist.push(res)
+          // tslint:disable-next-line:no-empty
           } catch (e) {}
```
Comment (SamRaymer) [inline src/rrule.ts]: oh my... i guess this is fine for now, but, oh my...

File: src/rrule.ts [hunk 7]
```diff
@@ -1167,22 +1353,23 @@ class Iterinfo {
     }
   }
 
-  ydayset (year, month, day) {
+  ydayset (_: any, __: any, ___: any) {
```
Comment (SamRaymer) [inline src/rrule.ts]: Can we just declare this as `ydayset () {` and get the same thing?

File: src/rrule.ts [hunk 8]
```diff
@@ -1167,22 +1353,23 @@ class Iterinfo {
...
     return [range(this.yearlen), 0, this.yearlen]
   }
 
-  mdayset (year, month, day) {
+  mdayset (_: any, month: number, __: any) {
```
Comment (SamRaymer) [inline src/rrule.ts]: Only the first two args seem useful here.

File: src/rrulestr.ts [hunk 9]
```diff
@@ -252,72 +347,54 @@ class RRuleStr {
...
-  'WE': 2,
-  'TH': 3,
-  'FR': 4,
-  'SA': 5,
-  'SU': 6
-}
 
-RRuleStr.prototype._handle_DTSTART = function (rrkwargs, name, value, options) {
-  rrkwargs[name.toLowerCase()] = dateutil.untilStringToDate(value)
+  // tslint:disable-next-line:variable-name
```
Comment (SamRaymer) [inline src/rrulestr.ts]: Maybe do a `// tslint:disable:variable-name` here and `// tslint:enable:variable-name` after this block of stuff

File: src/callbackiterresult.ts [hunk 10]
```diff
@@ -1,26 +1,33 @@
...
+  constructor (
+    method: string,
+    args: Partial<IterArgs>,
+    iterator: (d: Date, len: number) => boolean
+  ) {
     const allowedMethods = ['all', 'between']
     if (!contains(allowedMethods, method)) {
-      throw new Error('Invalid method "' + method +
-        '". Only all and between works with iterator.')
+      throw new Error(
```
Comment (SamRaymer) [inline src/callbackiterresult.ts]: 💛 optional: template string here

File: src/rrule.ts [hunk 11]
```diff
@@ -999,14 +1166,14 @@ class Iterinfo {
     const l = Math.floor(32 + 2 * e + 2 * i - h - k) % 7
     const m = Math.floor((a + 11 * h + 22 * l) / 451)
     const month = Math.floor((h + l - 7 * m + 114) / 31)
-    const day = (h + l - 7 * m + 114) % 31 + 1
+    const day = ((h + l - 7 * m + 114) % 31) + 1
```
Comment (SamRaymer) [inline src/rrule.ts]: 👍 Operator precedence is a little too magical sometimes, this is probably good.

File: src/rrule.ts (hunk 5 above)
Comment (SamRaymer) [reply src/rrule.ts]: Didn't realize there were like 15 of these, maybe not a good use of time after all

File: src/callbackiterresult.ts (hunk 10 above)
Comment (davidgoli) [reply src/callbackiterresult.ts]: I scrapped that bit altogether. Thanks, Typescript!

**Comment by davidgoli** [comment]
@arolson101 @jakubroztocil @thefliik I believe this branch is now ready for consideration. Please take a look!

**Comment by arolson101** [comment]
I was holding off looking at it because of all the commits going in after you asked for review ;)
Could you fix the conflict?

**Comment by davidgoli** [comment]
@arolson101 Conflicts resolved! Unfortunately, that last merge to master didn't come with a test, so I'm trying to write one now, but that shouldn't block consideration of this PR. Thanks!

**Comment by davidgoli** [comment]
@thefliik I have plans to break down the logic later on. It's spaghetti.

@arolson101 This is ready for a look. I have 2 follow on branches after this!

**Comment by arolson101** [comment]
I don't have time to look it over, but as long as it passes the test I'm happy

**Pull request merged**